### PR TITLE
feat: api for blueprints to manage ab player sessions with partInstance awareness

### PR DESCRIPTION
**Pull request by Julusian**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Feature

* **What is the current behavior?** (You can also link to an open issue here)

Blueprints are left to manage ab playback sessions themselves.
This is not possible to do well for partInstances, as more knowledge about the partInstances needs to be known than we are exposing to the blueprints

* **What is the new behavior (if this is a feature change)?**

This adds an a couple of methods available to onTimelineGenerate to help with ensuring ab-session ids are unique. 
With this api, unique session ids will be used for each instance of a part. The sessionName is used as an identifier for sessions, and multiple pieces using the same sessionName will be treated as the same session, even across parts. This lets a part continue a clip from the previous part. The exception here is that if the playhead moves backwards, then the session is split (this allows for having the same part as current and next).

* **Other information**:

Blueprints-integration changes: https://github.com/nrkno/tv-automation-sofie-blueprints-integration/pull/82
Blueprints changes: https://github.com/nrkno/tv-automation-sofie-blueprints/pull/73

**Status**

- [x] The functionality has been tested by the PR author
- [ ] The functionality has been tested by NRK

### CODE REVIEW

File: meteor/server/api/playout/timeline.ts [hunk 1]
```diff
@@ -274,19 +283,19 @@ function getTimelineRundown(cache: CacheForRundownPlaylist, studio: Studio): Tim
...
 						})
 					})
-					if (tlGenRes.persistentState) {
-						cache.RundownPlaylists.update(playlist._id, {
-							$set: {
-								previousPersistentState: tlGenRes.persistentState,
-							},
-						})
-					}
+					console.log(`AB Session result: ${JSON.stringify(context.knownSessions)}`)
```
Comment (sparkpunkd) [inline meteor/server/api/playout/timeline.ts:285]: Not required in production.

**Review by sparkpunkd — APPROVED** [review]
Moving the state information about sessions into the TimelineContext in core is a good solution ... and I'm pleased that this could be done both by extending both the existing TimelineContext and preserving the AB session info in the database. :-) 

Also pleased to see some unit test coverage, including for some bad parameters. 

I'm approving this on the assumption that:
* AB playback will work for ABC and ABCD - I see no reason why it would not;
* persisting the AB session info will mean good decisions continue to be made in the event of a core restart.

**Comment by Julusian** [comment]
@sparkpunkd for your assumptions:
* I havent tested this implementation, but there is no reason it shouldnt work. This code in core doesn't have any context of time or assign anything to a player, so it is completely agnostic.
* The primary aim of persisting is to be consistent with assignments during normal playback. But yes, it will work equally well around a restart

File: meteor/client/ui/SegmentTimeline/SegmentContextMenu.tsx [hunk 2]
```diff
@@ -36,8 +36,8 @@ export const SegmentContextMenu = withTranslation()(
 			const timecode = this.getTimePosition()
 			const startsAt = this.getPartStartsAt()
 
-			const isCurrentPart =
-				(part && this.props.playlist && part.instance._id === this.props.playlist.currentPartInstanceId) || undefined
+			const isCurrentPart = false
+			// (part && this.props.playlist && part.instance._id === this.props.playlist.currentPartInstanceId) || undefined
```
Comment (Julusian) [inline meteor/client/ui/SegmentTimeline/SegmentContextMenu.tsx:40]: This line should be reverted before merging, as non-ab setups havent been resolved yet

**Pull request merged**